R/utilities-data-combined.R

```diff
@@ -1,4 +1,344 @@
+#' sim <- loadSimulation(simFilePath)
+#' simResults <- runSimulation(sim)
+#' outputPath <- "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
+#'
+#' # observed data
+#' obsData <- lapply(
+#'   c("ObsDataAciclovir_1.pkml", "ObsDataAciclovir_2.pkml", "ObsDataAciclovir_3.pkml"),
+#'   function(x) loadDataSetFromPKML(system.file("extdata", x, package = "ospsuite"))
+#' )
+#' names(obsData) <- lapply(obsData, function(x) x$name)
+#'
+#'
+#' # Create a new instance of `DataCombined` class
+#' myDataCombined <- DataCombined$new()
+#'
+#' # Add simulated results
+#' myDataCombined$addSimulationResults(
+#'   simulationResults = simResults,
+#'   quantitiesOrPaths = outputPath,
+#'   groups = "Aciclovir PVB"
+#' )
+#'
+#' # Add observed data set
+#' myDataCombined$addDataSets(obsData$`Vergin 1995.Iv`, groups = "Aciclovir PVB")
+#'
+#' calculateResiduals(myDataCombined)
+#' @export
+calculateResiduals <- function(dataCombined,
+                               xUnit = NULL,
+                               yUnit = NULL,
+                               scaling = tlf::Scaling$lin) {
+  .validateScalarDataCombined(dataCombined)
+
+  # Validation has already taken place in the calling plotting function
+  combinedData <- dataCombined$toDataFrame()
+
+  # Remove the observed and simulated datasets which can't be paired.
+  combinedData <- .removeUnpairableDatasets(combinedData)
+
+  # Return early if there are no pair-able datasets present
+  if (nrow(combinedData) == 0L) {
+    warning(messages$residualsCanNotBeComputed())
+    return(NULL)
+  }
+
+  # Getting all datasets to have the same units.
+  combinedData <- .unitConverter(combinedData, xUnit, yUnit)
+
+  # Create observed versus simulated paired data using interpolation for each
+  # grouping level and combine the resulting data frames in a row-wise manner.
+  #
+  # Both of these routines will be carried out by `dplyr::group_modify()`.
+  pairedData <- combinedData %>%
+    dplyr::group_by(group) %>%
+    dplyr::group_modify(.f = ~ .extractResidualsToTibble(.x, scaling)) %>%
+    dplyr::ungroup()
+
+  return(pairedData)
+}
+
+#' Created observed versus simulated paired data
+#'
+#' @param data A data frame from `DataCombined$toDataFrame()`, which has been
+#'   further tidied using `.removeUnpairableDatasets()` and then
+#'   `.unitConverter()` functions.
+#'
+#' @examples
+#' # create an example data frame
+#' df <- dplyr::tibble(
+#'   dataType = c(rep("observed", 5), rep("simulated", 3)),
+#'   xValues = c(1, 3, 3.5, 4, 5, 0, 2, 4),
+#'   xUnit = ospUnits$Time$min,
+#'   xDimension = ospDimensions$Time,
+#'   yValues = c(1.9, 6.1, 7, 8.2, 1, 0, 4, 8),
+#'   yErrorValues = rnorm(8),
+#'   yUnit = ospUnits$`Concentration [mass]`$`mg/l`,
+#'   yDimension = ospDimensions$`Concentration (mass)`
+#' )
+#'
+#' ospsuite:::.extractResidualsToTibble(df)
+#'
+#' @keywords internal
+.extractResidualsToTibble <- function(data, scaling = tlf::Scaling$lin) {
```

Comment:
I think the `scaling `argument should be **mandatory** here?

Reply:
Good idea. Done!

Reply:
hmm, it's still optional?

Reply:
Oh, this is internal function. I changed it in `calculateResiduals`, which is the exported function.

I guess I can also change it here.

Reply:
Done.